### ISSUE TITLE
logic bug in Pickchildren conditions? - very, very slow excecution times

### ISSUE THREAD

**Issue by ruskul** [Scirra/Construct-bugs#8230]
## Problem description

The execution time for picking children through conditions in the event sheet is very, very slow.  Around maybe 120 times slower than manually tracking and picking those same children through uid.

## Attach a .c3p

https://drive.google.com/file/d/1Aq1BkpicNGKQhw9GnLaNRa-KhgsLy94Z/view?usp=sharing

## Steps to reproduce

The project has some code blocks that can be enabled and disabled to compare.   I have instructions located in the file, and there is some discussion located at: https://www.construct.net/en/forum/construct-3/general-discussion-7/looping-though-children-183621#forumPost1112988

Basically, all you need to do is disable one block and enable the other.  When you run the project, holding down space bar causes the loop to run.

Method 1 uses object hierarchies to iterate through a bunch of objects.  Method 2 manually adds "children" to an array that is created with each parent, and then iterates through those parents to get the array, then iterates the array to pick children by uid/

## Observed result

If you look at performance, method 2 is faster by orders of consequential magnitude.  It is around 120 times faster on my device than method 1.  Even if you reduce the number of children an object has to 1, it is still very slow.

## Expected result

I expect that picking children of a parent object should, if anything, be faster than manually tracking their uids, and using an array to loop through them. 

## More details

**Affected browsers/platforms:** 

Chrome and Firefox both have the same issue.

**First affected release:** 

## System details

<details><summary>View details</summary>

Platform information
Product: Construct 3 r407 (stable)
Browser: Chrome 128.0.6613.138
Browser engine: Chromium
Context: webapp
Operating system: Windows 10
Device type: desktop
Device pixel ratio: 1.25
Logical CPU cores: 8
Approx. device memory: 8 GB
User agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/128.0.0.0 Safari/537.36
Language setting: en-US

Local storage
Storage quota (approx): 142 gb
Storage usage (approx): 73 mb (0.1%)
Persistant storage: No

Browser support notes
This list contains missing features that are not required, but could improve performance or user experience if supported.

Nothing is missing. Everything is OK!
WebGL information
Version string: WebGL 2.0 (OpenGL ES 3.0 Chromium)
Numeric version: 2
Supports NPOT textures: yes
Supports GPU profiling: no
Supports highp precision: yes
Vendor: Google Inc. (Intel)
Renderer: ANGLE (Intel, Intel(R) UHD Graphics 630 (0x00003E9B) Direct3D11 vs_5_0 ps_5_0, D3D11)
Major performance caveat: no
Maximum texture size: 16384
Point size range: 1 to 1024
Extensions:

EXT_clip_control
EXT_color_buffer_float
EXT_color_buffer_half_float
EXT_conservative_depth
EXT_depth_clamp
EXT_disjoint_timer_query_webgl2
EXT_float_blend
EXT_polygon_offset_clamp
EXT_render_snorm
EXT_texture_compression_bptc
EXT_texture_compression_rgtc
EXT_texture_filter_anisotropic
EXT_texture_mirror_clamp_to_edge
EXT_texture_norm16
KHR_parallel_shader_compile
NV_shader_noperspective_interpolation
OES_draw_buffers_indexed
OES_sample_variables
OES_shader_multisample_interpolation
OES_texture_float_linear
OVR_multiview2
WEBGL_blend_func_extended
WEBGL_clip_cull_distance
WEBGL_compressed_texture_s3tc
WEBGL_compressed_texture_s3tc_srgb
WEBGL_debug_renderer_info
WEBGL_debug_shaders
WEBGL_lose_context
WEBGL_multi_draw
WEBGL_polygon_mode
WEBGL_provoking_vertex
WEBGL_stencil_texturing
Audio information
System sample rate: 48000 Hz
Output channels: 2
Output interpretation: speakers
Supported decode formats:

WebM Opus (audio/webm;codecs=opus)
WebM Vorbis (audio/webm;codecs=vorbis)
MPEG-4 Opus (audio/mp4;codecs=opus)
MPEG-4 AAC (audio/mp4;codecs=mp4a.40.2)
MP3 (audio/mpeg)
FLAC (audio/flac)
PCM WAV (audio/wav;codecs=1)
Supported encode formats:

WebM Opus (audio/webm;codecs=opus)
MPEG-4 Opus (audio/mp4;codecs=opus)
MPEG-4 AAC (audio/mp4;codecs=mp4a.40.2)
Video information
Supported decode formats:

WebM AV1 (video/webm;codecs=av01.0.00M.08)
WebM VP9 (video/webm;codecs=vp9)
WebM VP8 (video/webm;codecs=vp8)
MPEG-4 AV1 (video/mp4;codecs=av01.0.00M.08)
MPEG-4 H.265 (video/mp4;codecs=hev1.1.2.L93.B0)
MPEG-4 H.264 (video/mp4;codecs=avc1.420034)
Supported encode formats:

WebM AV1 (video/webm;codecs=av01.0.00M.08)
WebM VP9 (video/webm;codecs=vp9)
WebM VP8 (video/webm;codecs=vp8)
WebM H.264 (video/webm;codecs=avc1.420034)
MPEG-4 AV1 (video/mp4;codecs=av01.0.00M.08)
MPEG-4 VP9 (video/mp4;codecs=vp9)
MPEG-4 H.264 (video/mp4;codecs=avc1.420034)

</details>

**Comment by F3der1co** [comment]
It seems like it literally iterates over all picked instances and checks if it has a parent that is picked and is the correct object type.

**Comment by AshleyScirra** [comment]
Event blocks are a fundamentally different paradigm to JavaScript code - it's not always helpful to compare them directly, as the event engine has a far more complex job.

The problem is the condition "Parent: pick own children 'child'" starts with no 'child' instance picks. So, as is normal for conditions in the event system, it iterates all 'child' instances picking the ones that meet the condition, which in this case is to find the ones that are direct children of the parent.

The performance pitfall is then that you've put a 'For each' above that so it repeats it per parent instance, resulting in O(n^2) efficiency.

I guess this could be optimized, but it doesn't look like a real-world use case - the for-each loops are redundant. Is this something that comes up in a real-world project with well-designed events that don't redundantly repeat things?

**Comment by dop2000** [comment]
> I guess this could be optimized, but it doesn't look like a real-world use case - the for-each loops are redundant. Is this something that comes up in a real-world project with well-designed events that don't redundantly repeat things?

Here is a screenshot from our project. It's a part of z-ordering code, which is pretty complex because it sorts objects of different types and some objects are connected to each other. In this example I'm picking portable walls, which may have other items hanging on them (windows, posters, pictures), which, in turn, can have drawing canvases attached to them.

![image](https://github.com/user-attachments/assets/50de1e92-e107-4ef9-b714-078ba193f819)

**Comment by ruskul** [comment]
> Event blocks are a fundamentally different paradigm to JavaScript code - it's not always helpful to compare them directly, as the event engine has a far more complex job.

Please Ignore the JS blocks - they have nothing to do with this.  I added them in later and forgot to "save as a different file", but I thought I indicated clearly in comments that they aren't meant to compare to the events.  Sorry for the confusion.  I was just curious to compare the js against the other js methods, not against the events.

The methods being compared are the events.  In method one, scene graphs are used to select children.  In method 2, arrays , paired to parents by uid lookup, are used to store child uids.  Method 2 is fast.   Method 1 is slow.   The subsequent loop foreach child is not the issue, as if you get rid of the foreach child loop, performance is still about the same (around 2fps for me).  Likewise if if you do care about js, the api call to pick children is slower than the js call to pick by uid.

> The problem is the condition "Parent: pick own children 'child'" starts with no 'child' instance picked. So, as is normal for conditions in the event system, it iterates all 'child' instances picking the ones that meet the condition, which in this case is to find the ones that are direct children of the parent.

I don't understand though, "pickbyUID" also has no objects picked upfront, but it doesn't loop through every object, as is normal, looking for the one that meets the condition, right? -  it implements a smarter process for lookup that works well even with 10s of thousands of objects..   The whole reason to create a scene graph is so that you can traverse the scene graph at some point, and not having that data accessible in a performant manner makes no sense - Recreating a scene graph via events, that runs more performant than a built in system should not be possible at all. 

> The performance pitfall is then that you've put a 'For each' above that so it repeats it per parent instance, resulting in O(n^2) efficiency.

I kinda need to select each parent individually in order to communicate specific info to the children...  Are you sure the O(n^2) isn't internal side?

Method 2 does the same thing with no issue -  Or are you saying there is an alternative better way to select an individual parent and an individual child. 

What alternative is there?

AFAIK Such a thing must be done if each child instance needs to be picked independently.  Example:  Papa spaceship has some stats, and children components on that ship need to know those stats in order to do their actions, which may include iterating their children in turn as well.  When objects need to be picked in groups (like containers work), then you fundamentally have to do a foreach loop.  Since containers don't work with families, you need a way of pairing objects together.  Scene graphs are exactly the tool for pairing objects together, and is faster than manually linking uids.

> I guess this could be optimized, but it doesn't look like a real-world use case - the for-each loops are redundant. Is this something that comes up in a real-world project with well-designed events that don't redundantly repeat things?

Lol, I see what you did there.

Unless I misunderstand, What exactly is redundantly recurrently repeating here here?  And how is this not a real-world use case?  Whereever I need a specific picking of parent to children, I need to look up children.

For one thing, if those children are in a family, but of unknown types, and each of those children (or parents) have data pairs that need to be picked (such as arrays, or dictionaries) we already have a real world situation where we need to run a foreach.  As a matter of fact, for every reason you need a foreach loop for a parent is also a reason why you may need to iterate the children individually  at a time, with their parent individually picked.  So how is that redundant?, or do I misunderstand ? 

----Detailed Complex setup ----
Spaceship game where users design their own crafts, and others are dynamically generated at runtime.

Every spaceship part is comprised of the following families:
view, collider. state.  In order to select them together, we must connect them together, thus requiring some sort of foreach and picking loop.

states are then comprised of child components describing functionality.  Those children can have children as well, thus potentially requiring a loop to pick them individually.

Each spaceship part is assembled via scenegraphs to create a larger ship, with the top parent being the CommandCenter part.  

For a particular part, in this case a 2 barreled turret, we have the following parts: base, control, and 2 plasma weapons.  The plasma weapons are made up of a weaponplatform and a projectilelaucher.

Composing a new ship part is a matter of defining a new template, and can be done at runtime.  At runtime,  we need to select each CommandCenter, pick its children parts so the commandCenter can communicate to them.  If we follow the path down to the plasma weapons on the turret, they need to know the current ammo in the turret they belong to before firing, if such a command has been given.  This means that the  weapons must be picked together with their own turret (separately form other turrets).  Since any part can be anywhere, and dynamically integrated at runtime,  we need to iterate our ship parts.

**Comment by ruskul** [comment]
As an aside, using scene graphs to automatically update child objects works well as expected.  Behind the scenes children are selected efficiently and their positions updated quickly.  And it works much better than trying to do so through events, obviously. 

 I can write js a script to manually track and update child objects positions and it works barely and insignificantly faster.  As expected, as it strips away some of the hierarchies checks that make it useful to a variety of situations.  As it stands, I need to use hierarchies, but If they aren't optimized for traversing children, then I'll have to also recreate that structure in code or events as well with optimized graph traversal.

**Comment by AshleyScirra** [comment]
I've come up with an optimization for the next beta for specifically when 'Pick children' is used with the specified child object with all instances initially picked. This makes the test project significantly faster, with only a small performance overhead when holding space. I think that seems to solve the problem at hand, so closing this issue.

The condition still may be slow if prior conditions pick child instances. If previous conditions have for example picked thousands of random parent and child instances, then the condition has to filter the already-picked children according to which are children of the already-picked parents, which is quite a complicated filtering operation. It does try to do this efficiently, but in intensive cases it may still be slow for the same reason. However so long as 'Pick children' is the first condition that references the child object, then it should be efficient.

**Comment by ruskul** [comment]
> I've come up with an optimization for the next beta for specifically when 'Pick children' is used with the specified child object with all instances initially picked. This makes the test project significantly faster, with only a small performance overhead when holding space. I think that seems to solve the problem at hand, so closing this issue.
> 
> The condition still may be slow if prior conditions pick child instances. If previous conditions have for example picked thousands of random parent and child instances, then the condition has to filter the already-picked children according to which are children of the already-picked parents, which is quite a complicated filtering operation. It does try to do this efficiently, but in intensive cases it may still be slow for the same reason. However so long as 'Pick children' is the first condition that references the child object, then it should be efficient.

Ah, that makes a lot of sense.  Actually, that makes sense now why it was programmed the way it was.   I canʻt think of a good reason I personally would need to preselect large groups of children and then use their parents to itterate through them, but it is good to know that forces a longer check.

   I am curious though about the reverse though.  Does this optimization also include going the opposite direction. - I do often have children picked through a condition (such as collisions), where the parent isnʻt already selected and I need to get the parent. 

Basically getting children (and allchildren), getting parent (and all parents), quickly when you have one selected is also used frequently in similar situations.

**Comment by dop2000** [comment]
> This makes the test project significantly faster, with only a small performance overhead when holding space. I think that seems to solve the problem at hand, so closing this issue.

@AshleyScirra This is great! Could you please also fix this in the LTS version?

**Comment by F3der1co** [comment]
side note, in dop200 example I would assume the zIndex resort is the bottleneck though.

**Comment by dop2000** [comment]
Yeah, I know that these actions are not fast. In any case that was just one example, we have over 500 "pick child" and "pick parent" conditions in our project. Fixing them will probably improve the performance in many parts of our game.

**Comment by AshleyScirra** [comment]
It turns out the 'Pick parent' and 'Pick Nth child' conditions happen to use a similar code pattern where the same optimization can be applied, so I've also done that for the next beta.

**Comment by dop2000** [comment]
@AshleyScirra Will LTS version receive this update? Many developers will be stuck with it for the foreseeable future.

**Comment by ruskul** [comment]
> @AshleyScirra Will LTS version receive this update? Many developers will be stuck with it for the foreseeable future.

I'm hopeful about that too.

**Comment by AshleyScirra** [comment]
From the manual section on [release channels](https://www.construct.net/en/make-games/manuals/construct-3/tips-and-guides/release-channels):

> If you are wondering if any particular fix or change will be applied to an LTS release, the default answer is "no".

Further, these conditions have been in Construct for around 4 years IIRC, and this is the first time I recall anyone reporting a performance problem with them, so it would seem it's been working fine for most people most of the time.

**Comment by dop2000** [comment]
@AshleyScirra  

> this is the first time I recall anyone reporting a performance problem with them, so it would seem it's been working fine for most people most of the time.

The reason this issue hasn't been reported earlier is that no one expected a basic condition like "Pick child" to have such a significant performance impact. However, that doesn’t mean the problem didn’t exist. This is a critical performance improvement, and as I mentioned, many developers (including myself) will be relying on the LTS version for years to come. [Our game](https://store.steampowered.com/app/1658150/Moonstone_Island/) sometimes struggles to perform well on lower-end hardware like the Steam Deck, and even a small improvement, such as a 2-5 fps boost, could make a difference.

**Comment by ruskul** [comment]
Ashley, please put it in the LTS.  I need it.  Everybody else who uses child picking will experience reduced carbon emissions.  Save a tree or two.  Think of the children's electric bill and their FPS.   I don't want to have to port my sdkv1 addons, so I'm likely stuck with lts, and I need this performance improvement.  

If you need more reasons:

Furthermore, Pretty sure I offhand mentioned last year that hierarchies were too slow to use widely in a project, along with other critical areas.  I thought it was related to graphs updating positions and such and never thought to check into the actual logic to iterate through children picking. 

The advice at the time (your's) was to just make the project anyway and worry about finding optimization bottle-necks in the future if needed.    Now it is the future.  

Part of that search led me to the bug in Dictionaries, and Arrays.  Nobody ever complained about that either despite living 10 years with it, but it was brutally slow in specific types of project organization.  The fact is, basic components like this need to be as fast as possible and they are always the last place I even think to look for performance issues.  Game engines need to idealy be as lean as possible, and this fix should totally go into the LTS.  It isn't a major feature, change, etc... but a small fix, and the LTS was just introduced, implication there being obvious.

**Comment by AshleyScirra** [comment]
My default assumption is that it is not important enough to warrant the risk of accidentally breaking an LTS release (especially as this change has not yet even made it to a stable release so we have not yet achieved high confidence it doesn't affect anything else). Breaking an LTS release is something we should ideally never do as it defeats the purpose of LTS releases, and every change carries a risk of breakage, and therefore the bar is high for LTS updates: it must be a really critical issue. This was in place for 4 years before anyone noticed, the performance problem only arises with fairly specific combinations of events, and even with the affected combination of events the problem is only significant with large numbers of instances, and even then, you can update to the latest beta (and next stable) and have the optimization.

What would be more persuasive that it is necessary to port to LTS is lots of people providing actual performance measurements with real-world projects, along the lines of "my project could only get 45 FPS before but with this change it gets 60 FPS". If lots of people are saying that kind of thing, then it may raise the apparent severity of the issue to be critical enough to worth running the risk of breaking LTS, and even then only some time after the next stable release. I'm afraid comments merely stating "in my opinion this should be in the LTS release" will not change the balance of the judgement unless they are backed by convincing measurements - as ever, just talking about performance is pretty meaningless, only measurements matter - and sufficient evidence it meets the "critical" threshold. These are the criteria we will be applying to any change to LTS releases.
